File: api/src/main/java/org/tailkeep/api/model/Video.java
package org.tailkeep.api.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(name = "video")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Video {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id;
    
    @Column(name = "youtube_id", unique = true, nullable = false)
    private String youtubeId;
    
    @ManyToOne(optional = false)
    @JoinColumn(name = "channel_id", nullable = false)
    private Channel channel;
    
    @Column(nullable = false)
    private String url;
    
    @Column(nullable = false)
    private String title;
    
    @Column(name = "duration_string")
    private String durationString;
    
    private Double duration;
    
    @Column(name = "thumbnail_url")
    private String thumbnailUrl;
    
    @Column(columnDefinition = "TEXT")
    private String description;
    
    @Column(name = "view_count")
    private Long viewCount;
    
    @Column(name = "comment_count")
    private Long commentCount;
    
    @Column(nullable = false)
    private String filename;
    
    @OneToMany(mappedBy = "video")
    private List<DownloadProgress> progressList;
    
    @Column(name = "created_at")
    @CreationTimestamp
    private LocalDateTime createdAt;
    
    @Column(name = "updated_at")
    @UpdateTimestamp
    private LocalDateTime updatedAt;
}

File: web/src/schemas/video.ts
import { z } from 'zod';
import { channelDtoSchema } from './channel';

export const videoSchema = z.object({
  youtubeId: z.string(),
  url: z.string(),
  title: z.string(),
  channel: channelDtoSchema,
  durationString: z.string(),
  duration: z.number(),
  thumbnailUrl: z.string(),
  description: z.string(),
  viewCount: z.number(),
  commentCount: z.number(),
  filename: z.string()
});

export type Video = z.infer<typeof videoSchema>;

export const videoDtoSchema = videoSchema.extend({
  id: z.string(),
  createdAt: z.coerce.date(),
  updatedAt: z.coerce.date()
});

export type VideoDto = z.infer<typeof videoDtoSchema>;
